pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,615 @@
+                            if (throwable != null && log.isDebugEnabled()) {
+                                log.debug("[{}] Failed to update index snapshot", topic.getName(), throwable);
+                            }
+                            STATE_UPDATER.compareAndSet(this, OperationState.Operating, OperationState.None);
+                        });
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }
+                case WriteSegment, DeleteSegment -> {
+                    if (!STATE_UPDATER.get(this).equals(OperationState.Closed)) {
+                        taskQueue.add(new MutablePair<>(operationType, new MutablePair<>(taskExecutedResult, task)));
+                        executeTask();
+                        return taskExecutedResult;
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }
+                //If the operation type is clear, all the operation in the queue is meaningless.
+                case Clear -> {
+                    //Do not clear the snapshots if the topic is used.
+                    if (STATE_UPDATER.compareAndSet(this, OperationState.None, OperationState.Closed)) {
+                        taskQueue.forEach(pair ->
+                                pair.getRight().getRight().get().completeExceptionally(
+                                        new BrokerServiceException.TransactionBufferClosedException(
+                                                String.format("Cancel the operation [%s] due to the"
+                                                                + " transaction buffer of the topic[%s] already closed",
+                                                        pair.getLeft().name(), this.topic.getName()))));
+                        taskQueue.clear();
+                        return task.get();
+                    } else {
+                        return FutureUtil.failedFuture(
+                                new BrokerServiceException.NotAllowedException(
+                                        String.format("Failed to clear the snapshot of topic [%s] due to "
+                                                + "the topic is used. Please stop the using of the topic "
+                                                + "and try it again", this.topic.getName())));
+                    }
+                }
+                default -> {
+                    return FutureUtil.failedFuture(new BrokerServiceException
+                            .NotAllowedException(String.format("Th operation [%s] is unsupported",
+                            operationType.name())));
+                }
+            }
+        }
+
+        private void executeTask() {
+            if (taskQueue.isEmpty()) return;
+            OperationType operationType = taskQueue.getFirst().getKey();
```

Comment:
still will throw the NullPointException, two thread will get the firstKey.
appendTask thread and                     topic.getBrokerService().getPulsar().getTransactionExecutorProvider().getExecutor(this)